### ISSUE TITLE
Grpc contract method parameter count limited to 3 ?

### ISSUE THREAD

**Issue by menaheme** [protobuf-net/protobuf-net.Grpc#191]
I am reading through some of the code and noticed this [limitation](https://github.com/protobuf-net/protobuf-net.Grpc/blob/18ddc4268af6ed7c77c60a8fc67a7d241f6c95ee/src/protobuf-net.Grpc/Internal/ContractOperation.cs#L212)

of parameter count to 3, is this really necessary ? also, thers no log message to inform the poor programmer why the method was not bound.

Am i missing something here ?

**Comment by mgravell** [comment]
> Am i missing something here ?

Yes; gRPC is predicated on sending a single payload parameter, and as such you should *only ever really need one data parameter* - which might be a data type *with multiple members*. The other 2 parameters it accepts are so it can infer a `CancellationToken`, or context data. It is *not* intended for additional data parameters.

Now: this has confused a few people, and there *is* a PR and plan that changes it to take multiple data parameters (spoofing the underlying data parameter), but: this is incomplete and I haven't had time to visit it. 2020/2021 have been... hard.

**Comment by menaheme** [comment]
Thanks for the clarification.

Since the code first approach is based on .net interfaces, this restriction is less obvious.
The fact that you have already stared work on a PR is appreciated and respected. 
I humbly suggest a log entry form the build tools to be a temporary solution 

Indeed trying times, 
Regards from Israel

**Comment by mgravell** [comment]
I actually thought the build tools caught this one :shrug:

**Comment by menaheme** [comment]
Theres a PR that handles the parameter limit,  so im closing the issue
